### PR TITLE
CP-22597: Add vGPU-to-GPU-group map to SXM migration

### PR DESCRIPTION
**Pull request by mseri**
This PR also meant to address the potential issue discovered reviewing https://github.com/xapi-project/xen-api/pull/3058

However it looks like this was a non-issue, in the sense that a failure in calling `allocate_vm_to_host` during SXM migration would trigger the cleanup already in there, while a failure later on during the migration would anyway trigger the cleanup on the remote host that will take care of everything else.

This PR needs a small change to xapi-idl to introduce the other_config field to the vgpu interface, this is needed as temporary store to mimick the map functionality that we are already using for vifs and vbds.

### CODE REVIEW

**Comment by mseri** [comment]
Note: I have provoked SXM failures at different stages and it, indeed, seems that the cleanup is happening properly.

File: ocaml/idl/datamodel.ml [hunk 1]
```diff
@@ -2450,12 +2450,15 @@ let vm_migrate_send = call
...
-             Map (String, String), "options", "Other parameters"]
+    ~versioned_params:
+      [{param_type=Ref _vm; param_name="vm"; param_doc="The VM"; param_release=tampa_release; param_default=None};
+       {param_type=Map(String,String); param_name="dest"; param_doc="The result of a Host.migrate_receive call."; param_release=tampa_release;  param_default=None};
+       {param_type=Bool; param_name="live"; param_doc="Live migration"; param_release=tampa_release; param_default=None};
+       {param_type=Map (Ref _vdi, Ref _sr); param_name="vdi_map"; param_doc="Map of source VDI to destination SR"; param_release=tampa_release; param_default=None};
+       {param_type=Map (Ref _vif, Ref _network); param_name="vif_map"; param_doc="Map of source VIF to destination network"; param_release=tampa_release; param_default=None};
+       {param_type=Map (String, String); param_name="options"; param_doc="Other parameters"; param_release=tampa_release; param_default=None};
+       {param_type=Map (Ref _vgpu, Ref _gpu_group); param_name="vgpu_map"; param_doc="Map of source vGPU to destination GPU group"; param_release=vgpu_tech_preview_release; param_default=Some (VMap [])}
```
Comment (thomassa) [inline ocaml/idl/datamodel.ml]: `vgpu_tech_preview_release` was a tech-preview of the first ever vGPU functionality.
We shouldn't use it here. We should use the first release in which this new param will appear: one that has not yet been released.

File: ocaml/idl/datamodel.ml [hunk 2]
```diff
@@ -2465,13 +2468,15 @@ let vm_assert_can_migrate = call
...
-      Map (String, String), "options", "Other parameters" ]
+    ~versioned_params:
+      [{param_type=Ref _vm; param_name="vm"; param_doc="The VM"; param_release=tampa_release; param_default=None};
+       {param_type=Map(String,String); param_name="dest"; param_doc="The result of a VM.migrate_receive call."; param_release=tampa_release;  param_default=None};
+       {param_type=Bool; param_name="live"; param_doc="Live migration"; param_release=tampa_release; param_default=None};
+       {param_type=Map (Ref _vdi, Ref _sr); param_name="vdi_map"; param_doc="Map of source VDI to destination SR"; param_release=tampa_release; param_default=None};
+       {param_type=Map (Ref _vif, Ref _network); param_name="vif_map"; param_doc="Map of source VIF to destination network"; param_release=tampa_release; param_default=None};
+       {param_type=Map (String, String); param_name="options"; param_doc="Other parameters"; param_release=tampa_release; param_default=None};
+       {param_type=Map (Ref _vgpu, Ref _gpu_group); param_name="vgpu_map"; param_doc="Map of source vGPU to destination GPU group"; param_release=vgpu_tech_preview_release; param_default=Some (VMap [])}
```
Comment (thomassa) [inline ocaml/idl/datamodel.ml]: As above, this should be an upcoming release, not that old one.

Could this list of versioned_params be defined in a single place then shared between "migrate_send" and "assert_can_migrate"? (But that refactoring would be a separate commit, and possibly a separate pull-request etc.)

File: ocaml/xapi/cli_operations.ml [hunk 3]
```diff
@@ -2794,6 +2794,11 @@ let vm_migrate printer rpc session_id params =
              let vdi = Client.VDI.get_by_uuid rpc session_id vdi_uuid in
              let sr = Client.SR.get_by_uuid remote_rpc remote_session sr_uuid in
              vdi,sr) (read_map_params "vdi" params) in
+         
+         let vgpu_map = List.map (fun (vgpu_uuid,gpu_group_uuid) ->
+             let vdi = Client.VGPU.get_by_uuid rpc session_id vgpu_uuid in
```
Comment (thomassa) [inline ocaml/xapi/cli_operations.ml]: The variable-names `vdi` and `sr` are horribly confusing.
They should be `vgpu` and `gpu_group` or similar (or avoid naming them at all).

File: ocaml/xapi/import.ml [hunk 4]
```diff
@@ -1205,16 +1205,31 @@ module VGPU : HandlerTools = struct
       let vm = log_reraise
           ("Failed to find VGPU's VM: " ^ (Ref.string_of vgpu_record.API.vGPU_VM))
           (lookup vgpu_record.API.vGPU_VM) state.table in
-      let group = log_reraise
+      let group =
+        (* If we find the cross-pool migration key, attach the vgpu to the provided gpu_group... *)
+        if List.mem_assoc Constants.storage_migrate_vgpu_map_key vgpu_record.API.vGPU_other_config
+        then Ref.of_string (List.assoc Constants.storage_migrate_vgpu_map_key vgpu_record.API.vGPU_other_config)
```
Comment (thomassa) [inline ocaml/xapi/import.ml]: Do we need to think about what will happen if we get a value from other_config but it turns out not to be a valid vgpu_group ref?

File: ocaml/xapi/cli_operations.ml (hunk 3 above)
Comment (mseri) [inline ocaml/xapi/cli_operations.ml]: Well spotted, these are plain wrong! I must have deleted the commit while fixing a rebase issue. Thanks!

File: ocaml/xapi/import.ml (hunk 4 above)
Comment (mseri) [inline ocaml/xapi/import.ml]: It will fail the compatibility check called during `migrate_send` and `assert_can_migrate` and raise an error there, before starting the `import`

File: ocaml/idl/datamodel.ml (hunk 2 above)
Comment (mseri) [inline ocaml/idl/datamodel.ml]: I think we cannot, or we could only for a part of it: `dest` has a different meaning in the two calls, and the list of parameter is order-sensitive.

File: ocaml/idl/datamodel.ml (hunk 1 above)
Comment (mseri) [inline ocaml/idl/datamodel.ml]: Ops, my bad. Do you have a suggestion for what I should use?

File: ocaml/xapi/xapi_pgpu_helpers.ml [hunk 5]
```diff
@@ -181,24 +186,38 @@ let assert_destination_pgpu_is_compatible_with_vm ~__context ~vm ~host ?remote (
...
-            raise Api_errors.(Server_error (nvidia_tools_error, [Ref.string_of host]))
-        in 
-        Xapi_gpumon.Nvidia.assert_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~dest_host:host ~encoded_pgpu_metadata:pgpu_metadata
-    ) vgpus
+      | `nvidia -> 
+        let pgpu =
+          get_pgpus_in_gpu_group gpu_group
+          |> get_first_pgpu_of_implementation vgpu `nvidia
+        in test_nvidia_compatibility vgpu pgpu;
+      ) vgpus
```
Comment (thomassa) [inline ocaml/xapi/xapi_pgpu_helpers.ml]: It looks to me as if some refactoring could be worthwhile here: the two uses of `List.iter` are almost identical, differing only in that one has `get_pgpus_of_host host` where the other has `get_pgpus_in_gpu_group gpu_group`.

Maybe something like
```
List.iter (fun (vgpu, candidate_gpus) ->
  ...
  | `nvidia -> let pgpu = candidate_gpus |> get_first_pgpu_of_implementation vgpu `nvidia
    ...
)
```
with `candidate_gpus` coming from either
`get_pgpus_of_host host` or ` get_pgpus_in_gpu_group gpu_group`
(or maybe use `get_candidate_gpus: unit -> gpu list`)

File: ocaml/xapi/xapi_pgpu_helpers.ml [hunk 6]
```diff
@@ -151,11 +151,16 @@ let assert_destination_pgpu_is_compatible_with_vm ~__context ~vm ~host ?remote (
     | None -> Db.PGPU.get_supported_VGPU_types ~__context ~self:pgpu
     | Some (rpc, session_id) -> XenAPI.PGPU.get_supported_VGPU_types rpc session_id pgpu
   in
-  let get_pgpus host =
+  let get_pgpus_of_host host =
```
Comment (thomassa) [inline ocaml/xapi/xapi_pgpu_helpers.ml]: For consistency, I'd rather see the two `get_pgpus_...` function names have the same choice of `in` and `of`.

File: ocaml/xapi/import.ml (hunk 4 above)
Comment (thomassa) [inline ocaml/xapi/import.ml]: Right, I saw that when I'd read some more. This seems fine.

**Comment by mseri** [comment]
Good point. I'll rename the commit messages after I have addressed your changes, before squashing for merging

File: ocaml/xapi/xapi_pgpu_helpers.ml (hunk 6 above)
Comment (mseri) [inline ocaml/xapi/xapi_pgpu_helpers.ml]: I opted for `get_pgpus_of_gpu_group`

**Comment by thomassa** [comment]
The three new commits look good.

**Comment by mseri** [comment]
Squashed changes, merging
